### PR TITLE
allow advi on custom subnetworks

### PR DESCRIPTION
**Pull request by gianlucadetommaso**
Similarly to what already available for the Laplace approximation, we want to allow to run ADVI on any subsets of model parameters. This drastically increases the scalability of the method over large models.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

ADVI is necessarily run over all model parameters.

Issue Number: N/A

## What is the new behavior?

The user can configure a `which_params` argument in the `ADVIPosteriorApproximator`, including a tuple of paths (i.e. list of strings) pointing to the parameter that ADVI will be applied upon.

### CODE REVIEW

**Comment by albertogaspar** [comment]
A general comment on commits: squash them into a single one and provide the details in the message.

File: fortuna/prob_model/posterior/base.py [hunk 1]
```diff
@@ -145,3 +153,58 @@ def save_state(
             checkpoint_path=checkpoint_path,
             keep=keep_top_n_checkpoints,
         )
+
+    def _sample_diag_gaussian(self, rng: Optional[PRNGKeyArray] = None, **kwargs) -> JointState:
```
Comment (albertogaspar) [inline fortuna/prob_model/posterior/base.py:157]: Maybe it's better to have this in a different class then the abstract Posterior. Only Laplace and ADVI need it for now right?

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_approximator.py [hunk 2]
```diff
@@ -22,10 +24,21 @@ def __init__(
...
+        which_params: Optional[Tuple[List, ...]]
+            Sequences of keys to the parameters of the probabilistic model for which to define the Laplace
+            approximation. If `which_params` is not available, the posterior approximation will be over all parameters.
         """
         self.std_init_params = std_init_params
         self.std_base = std_base
         self.n_loss_samples = n_loss_samples
 
+        if which_params:
+            if type(which_params) != tuple:
```
Comment (albertogaspar) [inline fortuna/prob_model/posterior/normalizing_flow/advi/advi_approximator.py:36]: It's better to use `isinstance` for type comparison as it can handle subclasses as well.

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py [hunk 3]
```diff
@@ -74,11 +79,71 @@ def fit(
...
+        allowed_states = [MAPState, ADVIState]
+        status = dict()
+
+        if fit_config.checkpointer.restore_checkpoint_path is not None:
+            state = self.restore_checkpoint(
+                restore_checkpoint_path=fit_config.checkpointer.restore_checkpoint_path,
+                optimizer=fit_config.optimizer.method,
+            )
+
+            if type(state) not in allowed_states:
```
Comment (albertogaspar) [inline fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py:93]: Like above, use `isinstance`

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py [hunk 4]
```diff
@@ -74,11 +79,71 @@ def fit(
...
             )
 
-        init_prob_model_state, n_train_data, n_val_data = self._init(
-            train_data_loader, val_data_loader
+        state = None
+        stds = None
+        allowed_states = [MAPState, ADVIState]
+        status = dict()
+
+        if fit_config.checkpointer.restore_checkpoint_path is not None:
```
Comment (albertogaspar) [inline fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py:87]: What do you think about extracting a method for this part? It would improve readability

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py [hunk 5]
```diff
@@ -63,6 +67,7 @@ def fit(
         self,
         train_data_loader: DataLoader,
         val_data_loader: Optional[DataLoader] = None,
+        map_fit_config: Optional[FitConfig] = None,
```
Comment (albertogaspar) [inline fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py:70]: I'd rather have a user explicitly train a Model using MAP and then training with ADVi on top by either passing the MAP state or the checkpoint. I think that having `map_fit_config` and training MAP within ADVI adds an unnecessary overhead.

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_trainer.py [hunk 6]
```diff
@@ -1,8 +1,65 @@
...
+            return FrozenDict(all_params)
+
+        params = self._unravel(mean_logvar_rav["mean"]).unfreeze()
+        stds = self._unravel(jnp.exp(0.5 * mean_logvar_rav["logvar"]))
+        for k, v in params.items():
+            params[k] = {"params": dict(mean=v["params"], std=stds[k]["params"])}
+        return FrozenDict(params)
+
+    def on_train_end(self, state: ADVIState) -> ADVIState:
+        self.save_checkpoint(
```
Comment (albertogaspar) [inline fortuna/prob_model/posterior/normalizing_flow/advi/advi_trainer.py:59]: You can call `super` here

File: fortuna/prob_model/posterior/normalizing_flow/normalizing_flow_trainer.py [hunk 7]
```diff
@@ -232,3 +209,14 @@ def validation_step(
                 **{f"val_{m}": v for m, v in val_metrics.items()},
             }
         return dict(val_loss=loss)
+
+    def _get_params_from_rav(self, _rav, unravel) -> Params:
+        if self._which_params is None:
+            return unravel(_rav)
+        return FrozenDict(
+            nested_set(
```
Comment (albertogaspar) [inline fortuna/prob_model/posterior/normalizing_flow/normalizing_flow_trainer.py:217]: [nit] can you specify the name of the arguments when you have a function call with multiple args? It greatly improved readability and is safer.

**Review by albertogaspar — CHANGES_REQUESTED** [review]

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py (hunk 4 above)
Comment (gianlucadetommaso) [reply fortuna/prob_model/posterior/normalizing_flow/advi/advi_posterior.py:87]: Yes, I'll do that.

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_trainer.py (hunk 6 above)
Comment (gianlucadetommaso) [reply fortuna/prob_model/posterior/normalizing_flow/advi/advi_trainer.py:59]: I over-wrote also `save_checkpoint`, so I think here it should be `self`.

File: fortuna/prob_model/posterior/base.py (hunk 1 above)
Comment (albertogaspar) [reply fortuna/prob_model/posterior/base.py:157]: Sounds good :)

File: fortuna/prob_model/posterior/normalizing_flow/advi/advi_trainer.py (hunk 6 above)
Comment (albertogaspar) [reply fortuna/prob_model/posterior/normalizing_flow/advi/advi_trainer.py:59]: No you can call `super`, the `self.save_checkpoint` will be on the `ADVITrainer`
